src/components/FormElements/AssetPicker.tsx

```diff
@@ -0,0 +1,311 @@
+  includeNative = true,
+  onChange,
+  labelSuffix,
+  fitContent,
+}: AssetPickerProps) => {
+  const initErrorState = {
+    code: "",
+    issuer: "",
+  };
+
+  const getInitialValue = () => {
+    if (variant === "string") {
+      const assetString = value?.split(":");
+      return {
+        type: undefined,
+        code: assetString?.[0] ?? "",
+        issuer: assetString?.[1] ?? "",
+      };
+    }
+
+    return {
+      type: value?.type,
+      code: value?.code ?? "",
+      issuer: value?.issuer ?? "",
+    };
+  };
+
+  const [stateValue, setStateValue] = useState(getInitialValue());
+  const [error, setError] = useState(initErrorState);
+
+  let stringOptions: AssetString[] = [
+    {
+      id: "native",
+      label: "Native",
+      value: "native",
+    },
+    {
+      id: "issued",
+      label: "Issued",
+      value: "",
+    },
+  ];
+
+  if (includeNone) {
+    stringOptions = [
+      {
+        id: "none",
+        label: "None",
+        value: "",
+      },
+      ...stringOptions,
+    ];
+  }
+
+  let objectOptions: AssetObject[] = [
+    {
+      id: "credit_alphanum4",
+      label: "Alphanumeric 4",
+      value: {
+        type: "credit_alphanum4",
+        code: "",
+        issuer: "",
+      },
+    },
+    {
+      id: "credit_alphanum12",
+      label: "Alphanumeric 12",
+      value: {
+        type: "credit_alphanum12",
+        code: "",
+        issuer: "",
+      },
+    },
+    // TODO: add Liquidity Pool shares (for Change Trust operation)
+  ];
+
+  if (includeNative) {
+    objectOptions = [
+      {
+        id: "native",
+        label: "Native",
+        value: {
+          type: "native",
+          code: "",
+          issuer: "",
+        },
+      },
+      ...objectOptions,
+    ];
+  }
+
+  // Extra helper function to make TypeScript happy to get the right type
+  const handleOnChange = (
+    id: AssetType | undefined,
+    value: string | AssetObjectValue | undefined,
+  ) => {
+    if (!value) {
+      onChange(id, undefined);
+    }
+
+    if (variant === "string") {
+      onChange(id, value as string);
+    } else {
+      onChange(id, value as AssetObjectValue);
+    }
+  };
+
+  const handleOptionChange = (
+    optionId: AssetType | undefined,
+    optionValue: string | AssetObjectValue | undefined,
+  ) => {
+    handleOnChange(optionId, optionValue);
+    setStateValue({ type: optionId, code: "", issuer: "" });
+    setError(initErrorState);
+  };
+
+  const validateCode = (code: string, assetType: AssetType | undefined) => {
+    if (!code) {
+      return "Asset code is required.";
+    }
+
+    let minLength;
+    let maxLength;
+
+    switch (assetType) {
+      case "credit_alphanum4":
+        minLength = 1;
+        maxLength = 4;
+        break;
+      case "credit_alphanum12":
+        minLength = 5;
+        maxLength = 12;
+        break;
+      default:
+        minLength = 1;
+        maxLength = 12;
+    }
+
+    if (!code.match(/^[a-zA-Z0-9]+$/g)) {
```

Comment:
do you think this regex will be re-used?

Reply:
Can't think of any other use cases right now. :thinking: If it is used anywhere else, we'll make it a const. :+1: